[WIP] VECM

yogabonito opened the pull request:
Continuing the work from #3070.

ChadFulton commented:
Just wanted to let you know that I am going to make an effort to help review this.

yogabonito commented:
Thank you @ChadFulton! I had an exam this week and will now finish the integration of Josef's code (cointegration test) ASAP. IMO the PR will be ready to be merged once this is done.

yogabonito commented:
Travis is giving a `FileNotFoundError` while in AppVeyor all tests pass. @ChadFulton, may I ask you to have a look at it?

yogabonito commented:
@josef-pkt I think that - apart from one open question - the code is ready to be merged. This question is related to the [`coint_trend` argument](https://github.com/statsmodels/statsmodels/pull/453/files#diff-feda438e36b7a1872c28d07dd824a7c9R119) of the `coint_johansen` function. This argument is not used in your tests and I haven't seen it in the corresponding [Matlab code](https://github.com/statsmodels/statsmodels/pull/453/files#diff-feda438e36b7a1872c28d07dd824a7c9R2). How shall this argument be treated?

josef-pkt commented:
AFAIR, I added it when a user asked for more flexible trend options. I only realized later that the options for the cointegration tests need to be limited to the cases for which we have the p-values or critical values for the test. The estimation of VECM can have more options, but the cointegration test can only allow the restricted set of options with p-values. We might be able to extend those a bit, but some combinations don't have a good (IIRC parameter or data independent) limiting distribution.

josef-pkt commented:
From what I can see in `setup.py` txt and csv files are only included in the distribution if the directory ends with `results`. I don't know why appveyor would pick them up anyway.

It looks like travis is using `python setup.py install` and appveyor uses `python setup.py develop`. So only travis is testing that the sdist includes all required files.

josef-pkt commented:
I don't know why travis doesn't find the csv file

One local check is to run `python setup.py sdist` and check in the created archive file if the csv/data files are included.

yogabonito commented:
Thanks for your help @josef-pkt! I really appreciate it!
The tar.gz-archive produced by `python3 setup.py sdist` contains the csv-file on my machine.

yogabonito commented:
@josef-pkt, now the tests are passing in Travis. To me it looks as if the AppVeyor build is failing randomly (this time a Python 2 build job failed, whereas after my last commit it was the Python 3 build job).

I think that the PR is now ready to be merged. What are the next steps? Shall I do a rebase using a copy of this branch (to avoid messing up this PR)?

josef-pkt commented:
If the rebase goes well, then force pushing on this branch is fine.
However, given the large number of commits, I would create a new branch for the rebase in case the rebase or merge conflicts is messy. If there are no conflicts, then force pushing here should be safe, otherwise open a new PR.

(There are two names in the previously rebased commits, but it looks fine in the network. It looks like the initial commits where with a different user name, but that wouldn't cause problems.)

josef-pkt commented on statsmodels/tsa/vecm/vecm.py:
```diff
@@ -0,0 +1,2146 @@
+        * -1 - no deterministic terms
+        * 0 - constant term
+        * 1 - linear trend
+        * >1 - higher polynomial order
```
are higher order polynomials supported, i.e. are the critical values for them?

yogabonito replied:
No, there are no critical values for them. But you have [tests for a `det_order` of 2](https://github.com/statsmodels/statsmodels/pull/453/files#diff-b8606f9b5bf6ab1cb94a6641dc1a2335R79). The functions which return critical values [return zeros](https://github.com/statsmodels/statsmodels/pull/453/files#diff-dddef066f83a428bdf10cafc9810e74aR92) in this case. Shall we raise an exception instead?

josef-pkt replied:
I don't see the context right now.
I think the convenience function for coint rank should raise an exception for all cases where we cannot determine the rank because we are missing the critical values.

The cointegration test should return nan for critical values and warn but not raise. In that case it can still be useful for the user to get the test statistic if the user has access to other tables, or wants to try bootstrap. (I think I have seen bootstrap articles, but didn't read any.)
Also the plan would be to add other approximation for the cointegration test critical/p values at some point in the future. (e.g. I looked at some other articles last summer like one or two by Doornik)

josef-pkt commented on statsmodels/tsa/vecm/vecm.py:
```diff
@@ -0,0 +1,2146 @@
+                "% significance level"
+        print(SimpleTable(data=verbose_data, headers=headers, title=title))
+
+    return r_0
```
I think this should return an instance of a simple Results class, the verbose part can go into `__str__` and the SimpleTable into `summary()`

josef-pkt commented on statsmodels/tsa/vecm/vecm.py:
```diff
@@ -0,0 +1,2146 @@
+    return r_0
+
+
+def coint_johansen(endog_tot, det_order, k_ar):
```
this is just a cleaned up version of my function, is it?

(AFAIU, we should be able to do get the same information from the VECM class.)

yogabonito replied:
I have basically moved your function to this place in order to have all VECM-related functionality inside the vecm-module. Maybe we should add an underscore to mark it as internal because the user will probably want to use `select_coint_rank` as this function is calling `coint_johansen` until the optimal rank has been found. What do you think?

josef-pkt replied:
I think we can keep coint_jophansen public when users want to see the full results. However, we could add a warning in the notes section that the implementation might change to use the VECM class.

josef-pkt commented on statsmodels/tsa/vecm/vecm.py:
```diff
@@ -0,0 +1,2146 @@
+
+    Parameters
+    ----------
+    endog_tot : array-like (nobs_tot x neqs)
```
just `endog`

josef-pkt commented on statsmodels/tsa/vecm/vecm.py:
```diff
@@ -0,0 +1,2146 @@
+    endog_tot : array-like (nobs_tot x neqs)
+        2-d endogenous response variable.
+    exog: ndarray (nobs_tot x neqs) or None
+        Deterministic terms outside the cointegration relation.
```
add "User provided deterministic terms and weakly exogenous variables ..."
add comment about `deterministic`, for example, "deterministic trends specified by the `deterministic` keyword will be added to it by the model"
That's what it does, does it?

josef-pkt replied:
also season dummies (?) will be added.

yogabonito replied:
`exog` is not affected by `deterministic` and `seasons`. But it is possible to get all the _estimators_ for the deterministic terms through the `det_coef` attribute of a `VECMResults` object. The parts of `det_coef` are also available via the `const`, `seasonal`, `lin_trend`, and `exog_coefs` attributes.

josef-pkt commented on statsmodels/tsa/vecm/vecm.py:
```diff
@@ -0,0 +1,2146 @@
+    References
+    ----------
+    .. [1] Lutkepohl, H. 2005. *New Introduction to Multiple Time Series Analysis*. Springer.
+    """
```
This needs a `Notes` section with more explanation about the options.
e.g. the exog, deterministic and seasons and their relationship could be explained here.

josef-pkt commented on statsmodels/tsa/vecm/vecm.py:
```diff
@@ -0,0 +1,2146 @@
+        See :class:`statsmodels.tsa.base.tsa_model.TimeSeriesModel` for more
+        information.
+    freq : str, optional
+        See :class:`statsmodels.tsa.base.tsa_model.TimeSeriesModel` for more
```
what is `freq` used for?
It should be the same as specifying season if it applies.
?

yogabonito replied:
I forgot `freq` and `missing` which were part of the scaffold I started from. After the last rebase I get warnings related to `freq`. Shall I just use `freq` instead of `seasons`?

josef-pkt commented on statsmodels/tsa/vecm/vecm.py:
```diff
@@ -0,0 +1,2146 @@
+        See :class:`statsmodels.tsa.base.tsa_model.TimeSeriesModel` for more
+        information.
+    missing : str, optional
+        See :class:`statsmodels.base.model.Model` for more information.
```
Is this really correctly handled for time series? Might need a warning or explanation, that dropping observations might not make sense.

yogabonito replied:
Shall I just remove `missing`?

yogabonito commented:
@josef-pkt, I have seen your suggestions right after force pushing. I will go through them now. Thanks for your feedback!
The two usernames stem from changing the git config during the development.

josef-pkt commented on statsmodels/tsa/vector_ar/irf.py:
```diff
@@ -277,7 +287,7 @@ def errband_mc(self, orth=False, svar=False, repl=1000,
                                         burn=burn, cum=False)
         else:
-            return model.irf_errband_mc(orth=orth, repl=repl, T=periods,
+            return model.irf_errband_mc(orth=orth, repl=repl, T=periods,  # TODO: irf_errband relies on self.intercept --> solve with exog-refactor
```
long comments (if beyond 80 character line limit) should go into separate line

josef-pkt commented on statsmodels/tsa/vector_ar/output.py:
```diff
@@ -173,9 +173,35 @@ def _resid_info(self):
-    null_hyp = 'H_0: %s do not Granger-cause %s' % (variables, equation)
+def causality_summary(results, variables, equation, kind, inst_caus=False):
+    """
+
```
one line summary missing (doc rendering will have problems)

josef-pkt commented:
@yogabonito Do you have any notebooks with for example the test cases against jmulti?
It would be good to add some to the documentation to illustrate the usage and features of VECM.

We don't commit the output in notebooks (to avoid increasing repo memory with rendered output and plots), so it's better to post draft versions somewhere else, e.g. in a gist or scratch/working repo on github.

yogabonito commented:
I will make a notebook after I have gone through your suggestions :)

yogabonito commented on statsmodels/tsa/vecm/vecm.py:
```diff
@@ -0,0 +1,2146 @@
+    """
+    u_, s_, v_ = svd(_2darray, full_matrices=False)
+    s_ = np.sqrt(s_)
+    return chain_dot(u_, np.diag(s_), v_)
```
`s_` is  a 1D-array, so using `*` won't be possible.

josef-pkt replied:
s_ should broadcasts. I use this shortcut all the time with diagonal matrices. I just need to check each time whether it needs to be row or column s_ or s_[:, None]

josef-pkt commented on statsmodels/tsa/vecm/vecm.py:
```diff
@@ -0,0 +1,2146 @@
+        if beta.dtype == np.complex128 and np.all(np.imag(beta) == 0):
+            beta = np.real(beta)
+        if gamma.dtype == np.complex128 and np.all(np.imag(gamma) == 0):
+            gamma = np.real(gamma)
```
numpy has a function `real_if_close` that drops noisy zero imaginary parts

josef-pkt replied:
When are those arrays, alpha, beta, gamma, complex?

yogabonito replied:
For some combinations of deterministic terms Gamma was returned with `dtype=complex` although the imaginary part was zero. (The dtype was changed by `np.inv()` somewhere in the code.) I have edited the code to incorporate your hint with `real_if_close`. Thanks for that!

josef-pkt commented on statsmodels/tsa/vecm/vecm.py:
```diff
@@ -0,0 +1,2146 @@
+        Estimate for the parameter :math:`\\alpha` of a VECM.
+    beta : ndarray (neqs x coint_rank)
+        Estimate for the parameter :math:`\\beta` of a VECM.
+    gamma : ndarray (neqs x neqs*(k_ar-1))
```
these three should be changed to something `params_xxx`

also we need informative names, and then a list of how they map to Luetkepohl's or Jorgensen's math notation

josef-pkt commented on statsmodels/tsa/vecm/vecm.py:
```diff
@@ -0,0 +1,2146 @@
+        attribute is an ndarray of size zero.
+    delta_y_1_T : see Parameters
+    y_min1 : see Parameters
+    delta_x : see Parameters
```
I think we make those 3 attributes private, unless we come up with good names, e.g. `_delta_y_1_T`

josef-pkt commented on statsmodels/tsa/vecm/vecm.py:
```diff
@@ -0,0 +1,2146 @@
+    delta_y_1_T : see Parameters
+    y_min1 : see Parameters
+    delta_x : see Parameters
+    r : int
```
use full name, e.g. `coint_rank`

josef-pkt commented on statsmodels/tsa/vecm/vecm.py:
```diff
@@ -0,0 +1,2146 @@
+        corresponding column of det_coef_coint. If there is no linear
+        deterministic term outside the cointegration relation, then
+        lin_trend is an ndarray of size zero.
+    exog_coefs : ndarray (neqs x exog_coefs.shape[1])
```
should also be `params_xxx`

josef-pkt commented on statsmodels/tsa/vecm/vecm.py:
```diff
@@ -0,0 +1,2146 @@
+    beta : see Parameters
+    gamma : see Parameters
+    sigma_u : see Parameters
+    det_coef_coint : ndarray (#(determinist. terms inside the coint. rel.) x r)
```
params_xxx

IIRC, we discussed somewhere about using a Params class so we can keep all the different components of params together.
right now I think it's enough if all parameters are clearly identifiable as such,  `params_xxx`

josef-pkt commented on statsmodels/tsa/vecm/vecm.py:
```diff
@@ -0,0 +1,2146 @@
+
+
+    @cache_readonly
+    def cov_params(self):  # p.296 (7.2.21)
```
I guess this will create a conflict with the inherited cov_params which is a method that also computes linear combinations.
maybe it's possible and enough to rename to cov_params_default.

josef-pkt commented:
One thing to check is inference. I didn't get so far yet.
We should get `t_test` to work. If it doesn't work, then we need to think about what changes we would need, even if they are not implemented before merging this PR. (There are other models that don't find the generic pattern in base where t_test is broken, IIRC)

josef-pkt commented on statsmodels/tsa/vecm/vecm.py:
```diff
@@ -0,0 +1,2146 @@
+                raise TypeError("causing has to be of type string or int (or "
+                                "a sequence of these types) or None.")
+            causing = [self.names[c] if type(c) == int else c for c in causing]
+            causing_ind = [get_index(self.name
```
use a results class and add `summary` method to it instead of using verbose.

josef-pkt replied:
this is in differences isn't it.
It should mention that it is the short run equation in first differences of endog

josef-pkt replied:
return instance of a results class instead of dict

josef-pkt commented:
I just remembered: 
Standard wald (t_test) inference for parameters is an issue for VAR and not specific to VECM, i.e. we want to test parameters for the VARX in differences for given cointegrating equations. So we will need to fix it later for VAR and is not required in the VECM PR. (VECM granger causality is implemented and doesn't use generic wald test code.)
Tests that also involve parameters from the cointegrating equations are a different issue and will eventually need special code for estimating VECM under parameter restrictions on `beta` or jointly on `alpha` and `beta`..

yogabonito commented:
Here's a [notebook](https://gist.github.com/yogabonito/b532e20b64587c19198f1aacb5ddfca2) showing the `vecm` module's functionality with the new results classes. (I have added them also to the `vector_ar` module and updated the test cases accordingly.)
I have changed the docstrings too but a few summaries still contain a line break because the sentence didn't quite fit into the line. Is this ok or do the summaries need to be shorter in these cases?
I will now go through your remaining suggestions :)

yogabonito commented:
@josef-pkt, I made the changes you suggested. Only `freq` and `missing` are unchanged. (I plan to remove the latter.)

I tried to make use of the `freq` argument instead of a `seasons` argument in `VECM`'s `__init__` method (not committed). However, there was a problem in the case of `endog` being an `ndarray` (`TimeSeriesModel`'s `_init_dates` is raising an exception) - see [this gist](https://gist.github.com/yogabonito/80a16b56feb849b29f78b55cdd0933fa).

So I think that it might be best to keep using a `seasons` argument in `VECM`'s `__init__` which can be used even if `endog` is an `ndarray`. It would also have the advantages
 * that it is numeric - as opposed to `freq` which is a string like "Q" or "W"
 * and that it can be set independently of `freq` (e.g. if the user wants to neglect the seasonality even if `freq` is specified).

What is your opinion?

Furthermore, what do you think about the current status of this PR? Do you have more suggestions for changes or can it be merged?

yogabonito commented:
@josef-pkt, @ChadFulton could you please have a look at the current state of this PR? What work is needed for this PR to get merged?

ChadFulton commented:
My apologies, I know I said I'd review this PR several months ago. It is on my list, but even so it may be several weeks before I can get back to it. I think it will be a great addition.

jbrockmendel commented on statsmodels/tsa/vector_ar/util.py:
```diff
@@ -193,7 +197,7 @@ def varsim(coefs, intercept, sig_u, steps=100, initvalues=None, seed=None):
     ugen = rmvnorm(np.zeros(len(sig_u)), sig_u, steps)
     result = np.zeros((steps, k))
-    result[p:] = intercept + ugen[p:]
+    result[p:] = intercept + ugen[p:] if intercept is not None else ugen[p:]
```
It isn't immediately obvious whether the order of operations is `(intercept+ugen[p:]) ...` or `intercept + (ugen[p:] ...)`.  In context it isn't hard to figure out, but I still suggest separating out the if/else condition.

This also helps make test coverage measurement more meaningful.

yogabonito replied:
OK, I understand. The corresponding change will be online with my next push.

jbrockmendel commented on statsmodels/tsa/vector_ar/tests/JMulTi_results/parse_jmulti_var_output.py:
```diff
@@ -0,0 +1,421 @@
+import re
```
There's a lot of work going on here to parse these results files.  Would it make sense to parse them once and commit them in a directly-importable form?

Related: the thoroughness is fantastic, but the large number of text files is daunting.  I imagine reviewers wouldn't be so reticent to jump in if it didn't say "208 files" up at the top of the window.

yogabonito replied:
Haha, I know that 208 files look like a lot of work for reviewing. Sorry for that! I thought that it would be the most transparent way to add the files generated by JMulTi directly to statsmodels. I will think about how I could accomplish this and appreciate any hints on that!
I am afraid that rewriting the program correspondingly would mean a lot of work too :smile: So if it's possible I would kindly ask you to ignore the high number of files. After all, the `.txt` files in `.../vector_ar/tests/JMulTi_results` and in `.../vecm/tests/JMulTi_results` constitute the majority of files and can safely be skipped when reviewing.

jbrockmendel commented on statsmodels/tsa/vector_ar/var_model.py:
```diff
@@ -1342,39 +1678,188 @@ def test_causality(self, equation, variables, kind='f', signif=0.05,
+                                    crit_value, pvalue, df, signif,
+                                    test="inst", method="wald")
+
+    def test_whiteness_new(self, nlags=10, signif=0.05, adjusted=False):
```
@josef-pkt This looks like what we've discussed in #3683.  It's almost identical to the version I whipped up this morning.

@yogabonito Do you have a good intuition for why the `k_ar` shows up in the degrees of freedom?  The Stata docs I found for this were also VAR-specific, but it is not clear to me why.  It seems like this kind of test should be applicable to more general time-series residuals.

yogabonito replied:
I have written the `test_whiteness_new` method because I realized that the VAR framework was lacking a formal test. (As discussed in #3683, the `test_whiteness` method is not a formal test.)
The `k_ar` stems from the corresponding formulas in section 4.4.3 in Lütkepohl (2005) which is about the Portmanteau Test. There, Lütkepohl sketches a proof of the limiting distribution and offers the reader to refer to Ahn (1988) for more details. As the Portmanteau test is described mainly for VARs and VECMs in Lütkepohl (2005), I am not quite sure how this test can be generalized to other models.

jbrockmendel commented on statsmodels/tsa/vector_ar/var_model.py:
```diff
@@ -1342,39 +1678,188 @@ def test_causality(self, equation, variables, kind='f', signif=0.05,
+        """
         Test white noise assumption. Sample (Y) autocorrelations are compared
-        with the standard :math:`2 / \sqrt{T}` bounds.
+        with the standard :math:`2 / \sqrt(T)` bounds.
```
I'm not too familiar with restructuredtext, have just been assuming that TeX would be close-enough.  Is that wrong?

yogabonito replied:
You're right! :smile: This will be changed back with my next push.

jbrockmendel commented on statsmodels/tsa/vector_ar/var_model.py:
```diff
@@ -734,14 +1001,24 @@ def forecast_interval(self, y, steps, alpha=0.05):
         forc_lower = point_forecast - q * sigma
         forc_upper = point_forecast + q * sigma
 
         return point_forecast, forc_lower, forc_upper
```
I put this in a PR that got lost somewhere, might as well fix it here.  The docstring has an incorrect order for the return variables.

yogabonito replied:
I will continue addressing your suggestions tomorrow. (I had some issues after an OS upgrade and now it's time to sleep.) Thanks @jbrockmendel for taking the time to review the changes in the code!

yogabonito replied:
Good point! Will be in my next push.

jbrockmendel replied:
Happy to help.  I'm looking forward to playing around with all this new functionality.

If you fix the docstring issue, you can add a note that it closes issue 3615.

yogabonito replied:
Thanks for the hint! I mentioned it in the commit message.

jbrockmendel commented on statsmodels/tsa/vector_ar/var_model.py:
```diff
@@ -308,6 +387,125 @@ def _reordered(self, order):
+    return mat([np.dot(coefs, P) for coefs in ma_mats])
+
+
+def test_normality(results, signif=0.05):
```
Is this used anywhere other than inside `VARResults`?  If not, why not put the content in that method? (I have the same question about `forecast` in the status-quo branch)

josef-pkt replied:
(just to this one, I will try to come back to this PR later)
Several of the diagnostic tests like normality and whiteness/autocorrelation are generically useful for multivariate data. Eventually we will need to collect them in a central location similar to the univariate diagnostics and specification tests. However, until we have those and know what we need for other cases, it's a good compromise to have them semi-independent of the usage in the current model.

yogabonito replied:
@jbrockmendel, yes, these functions are also used by the VECM-framework in `statsmodels/tsa/vecm/vecm.py`.

jbrockmendel commented on statsmodels/tsa/vector_ar/var_model.py:
```diff
@@ -308,6 +387,125 @@ def _reordered(self, order):
+    omni_dist = stats.chi2(results.neqs * 2)
+    omni_pvalue = float(omni_dist.sf(lam_omni))
+    crit_omni = float(omni_dist.ppf(1 - signif))
+
```
sm.stats.stattools has a `omni_normtest` and a `jarque_bera` test.  Can we make use of either of those here?  As above this is a pre-existing issue, not specific to this PR.

yogabonito replied:
I have checked the implementation in `sm.stats.stattools`. It relies on `scipy` and `scipy`'s code differs substantially from the test proposed in Lütkepohl (2005). To me it looks as if there exist several proposals of such a test in literature which are all based on the same asymptotics but differ nonetheless. I would suggest not to use `sm.stats.stattools` in this case because the reference software _JMulTi_ seems to rely on Lütkepohl's formulas. Changing to `sm.stats.stattools` would break the corresponding test cases (I tried it out).

jbrockmendel commented on statsmodels/tsa/vector_ar/irf.py:
```diff
@@ -343,7 +353,7 @@ def err_band_sz1(self, orth=False, svar=False, repl=1000,
         return lower, upper
 
-    def err_band_sz2(self, orth=False, repl=1000, signif=0.05,
+    def err_band_sz2(self, orth=False, svar=False, repl=1000, signif=0.05,
```
+1 on avoiding `NameError`s

jbrockmendel commented on statsmodels/tsa/vector_ar/var_model.py:
```diff
@@ -1342,39 +1678,188 @@ def test_causality(self, equation, variables, kind='f', signif=0.05,
+            u -= np.mean(u, axis=1).reshape((u.shape[0], 1))
+            result = np.zeros((self.neqs, self.neqs))
+            for t in range(lag, self.nobs):
+                result += u[:, t:t+1].dot(u[:, t-lag:t-lag+1].T)
```
Either I'm not understanding something here or there are many more dot-products being computed than are necessary.  Aside from de-meaning the residuals (which could be done once, outside the loop), couldn't we just call `_compute_acov`?

yogabonito replied:
Thanks for pointing this out! Now `_compute_acov` is used - not only in `var_model.py` but also in the corresponding method in `vecm.py`.

jbrockmendel commented on statsmodels/tsa/vector_ar/tests/test_var_jmulti.py:
```diff
@@ -0,0 +1,577 @@
+        load_data(ds, data)
+        results_ref[ds] = load_results_jmulti(ds, dt_s_list)
+        results_sm[ds] = load_results_statsmodels(ds)
+        return results_sm[ds], results_ref[ds]
```
Is the `return` here supposed to be a `yield`?

Anything that can be done to make these tests easier to copy/paste would be appreciated.

yogabonito replied:
Actually the `return` is unnecessary. The important task in the `setup` function is to fill the (global) `result_XXX` dictionaries.
In this case the return did no harm because we fill the dictionaries with the results for only _one_ dataset - so the `return` is not stopping the loop prematurely.
I will delete the `return` statement with the next commit. Thanks for making me think about it!

yogabonito replied:
> Anything that can be done to make these tests easier to copy/paste would be appreciated.

Not sure I quite understand what you mean. Where would you like to paste the tests?
As an explanation, when I wrote the tests I followed the pattern of [parametric tests](https://github.com/numpy/numpy/blob/master/doc/TESTS.rst.txt#parametric-tests) which uses `yield` statements inside `test_XXX` functions to generate tests with different parameters.

jbrockmendel replied:
The yielded tests are great, and the parameterization makes perfect sense given the thorough set of test cases you've put together.

Probably easiest just to describe exactly what I was having trouble doing.  The new `test_whiteness_new` method is a near-perfect match for #3683, so I wanted to load one of the test cases/results, poke at it interactively, and confirm that your version subsumes the half-baked version I threw together.  I had hoped to copy/paste the necessary code without having to checkout and set up the branch.

If I'm the only person interested in this type of use, then don't inconvenience yourself on my account.

yogabonito replied:
Did you manage to make use of the test cases? I guess that using the test cases without checking out my PR-branch will require you to copy `test_var_jmulti.py` as well as the directory `JMulTi_results`. Then, you can delete all `test_XXX` functions in `test_var_jmulti.py` you don't need. For testing your method you can then add a `test_XXX` function or change the existing `test_whiteness` function in `test_var_jmulti.py`. For running the tests quickly, I also have this Python-script on my local machine:
```
from statsmodels import api as sm
sm.tsa.var.test()
quit()
```
(The first ~50 test cases run by this script (with one slow test case) were already part of statsmodels before this PR and are defined in `test_var.py` and `test_svar.py`.)

If you also review the new VECM framework, I would recommend to check out the PR-branch since the tests there rely on a new dataset added to the `statsmodels/datasets` directory (called `interest_inflation`).

yogabonito commented:
Thanks for the replies and feedback! I will start to go through @jbrockmendel's comments now.

josef-pkt commented:
@yogabonito  I never managed to get seriously back into tsa this year.

I would like to merge this as is, and we can try it out in master, and make changes before the release if necessary. Also, we will consider it as experimental for one release cycle so we don't have strong restrictions on making changes as we get feedback or as we get more experience using it.

Can you rebase one more time, then I will merge.

Thanks, and congratulation to your pysal GSOC

yogabonito commented:
@josef-pkt this is great to hear, thanks! :grinning:
I will rebase tomorrow if you want to merge the code. But I think that there is still an [open question](https://github.com/statsmodels/statsmodels/pull/3246#issuecomment-293429304) regarding the `freq` argument. The current code relies on a `seasons` argument instead of `freq` for indicating seasonality. And not providing `freq` leads to `ValueWarning` since the last rebase. How shall we handle this?

josef-pkt commented:
Ok, I will check. (I'm flying to France tomorrow, so I might have to leave this for a few days.) 

statespace uses `seasonal` and `seasonal_periods` instead of `freq`. IIRC Chad and I had a discussion about this because `freq` is ambiguous or misleading, e.g.  #3758 (`seasonal_periods` is a bit long to type and used as attribute but not necessarily as keyword.)

In the worst case we will have to work with aliases if we need compatibility with different interfaces until those are made consistent.

josef-pkt commented:
I got confused, `freq` here are the time units for the date index, and not seasonal periods. AFAIU from Chads refactoring #3272, and your exception in the gist is that freq is inconsistent with using ndarrays because there is no datetime index. Your gist example should run if `freq` is None.
Any seasonal length in terms of number of periods per cycle needs to be independent of this.

The freq and time index is mainly used in forecasting to create the time index for forecast periods and to infer how many periods to forecast if a time range is specified. If this is a numpy array then the index/time is supposed to be just the indexing/arange.

josef-pkt commented:
I just realized after checking out the PR branch, that vecm is in its own directory with a single module instead of in `vector_ar`. There are two JMulTi_results folders, I guess one with the VECM and one with the VAR tests.
Moving vecm.py into vector_ar is a bit painful because github doesn't follow file moves or renames in the history display, but I think it is still better to move it.
The names in the JMulTi_results folders are all different so it should be possible to merge those two folders.

@yogabonito Do you see any problems with merging vecm folder into vector_ar folder?

yogabonito commented:
I don't think that it is a big problem to merge the two directories. It might be a bit confusing to have all the test output from JMulTi for both VAR and VECM in one directory, though.
I will now try to rebase first and then see how I can move the contents of vecm to vector_ar.

yogabonito commented:
@josef-pkt I have moved `vecm` to `vector_ar`. I have also rebased again today since there were a few new commits in upstream/master. Now this PR should be ready to get merged.

josef-pkt commented:
@yogabonito  Thanks, I will have another quick look, but expect to merge it by tomorrow, assuming all is green.  (Costa Brava is calling.:)

yogabonito commented:
All is green :grinning: Have a nice holiday, Josef!

josef-pkt commented:
@yogabonito  I'm merging this as is, move and rebase look good. 
This is a BIG contribution, and thanks for the GSOC.

Given the lack of reviewer time, there are still some details left to review and change. Followup issue is #3775 
One immediate question: Why does HypothesisTestResults and subclasses implement `__eq__`? What's the intention and purpose for this?

yogabonito commented:
> Why does HypothesisTestResults and subclasses implement `__eq__`?

Finding out whether two tests are testing the same thing probably isn't an indispensable feature so the `__eq__` method can safely be removed. I just felt like implementing it ;)

yogabonito commented:
It's great to see VECM merged! :smiley: Thank you @josef-pkt and @bashtage for leading me through the GSoC - it was a great experience!

yogabonito commented:
> future contributions always welcome!

@bashtage for now I will concentrate on my GSoC 2017 but it's definitely something I will consider!

jbrockmendel commented on statsmodels/tsa/vector_ar/vecm.py:
```diff
@@ -0,0 +1,2247 @@
+    .. [1] Lutkepohl, H. 2005. *New Introduction to Multiple Time Series Analysis*. Springer.
+    """
+
+    # todo: rewrite m such that a big (TxT) matrix is avoided
```
Can the nobs x nobs problem be avoided by separating out the left-multiplication by `delta_x.T`?:

```
not_m = inv(delta_x.dot(delta_x.T)).dot(delta_x) # shape --> (k_ar_diff*neqs, nobs)
pre_r0 = delta_y_1_T.dot(delta_x.T)              # shape --> (neqs, k_ar_diff*neqs)
pre_r1 = y_lag1.dot(delta_x.T)                   # shape --> (neqs, k_ar_diff*neqs)
r0 = delta_y_1_T - pre_r0.dot(not_m)
r1 = y_lag1 - pre_r1.dot(not_m)
```
